File: src/services/policy/engine.ts
import { z } from "zod"
import type { Rule } from "./store-sqlite"

const MINUTES_PER_HOUR = 60

export const SeverityEnum = z.enum(["ok", "warn", "crit"])
export const RoleEnum = z.enum(["admin", "manager", "operator"])

export const AlertSchema = z.object({
  id: z.string().min(1),
  kpiId: z.string().min(1),
  title: z.string().min(1),
  message: z.string().min(1),
  severity: SeverityEnum,
  delta: z.number().optional(),
})
export type Alert = z.infer<typeof AlertSchema>

export type Decision =
  | { type: "deny"; reason: string }
  | {
      type: "allow"
      execute: boolean
      needsApproval: boolean
      approverRoles?: string[]
      ruleId: string
      resolvedParams: Record<string, unknown>
    }

/**
 * Prüft ob die aktuelle Zeit innerhalb des definierten Zeitfensters liegt
 */
function withinWindow(
  win?: { days: number[]; start: string; end: string },
  now = new Date()
): boolean {
  if (win === undefined) {
    return true
  }
  const day = now.getDay() // 0..6 (0=Sunday)
  if (!win.days.includes(day)) {
    return false
  }
  const [sh, sm] = win.start.split(":").map(Number)
  const [eh, em] = win.end.split(":").map(Number)
  const t = now.getHours() * MINUTES_PER_HOUR + now.getMinutes()
  const s = sh * MINUTES_PER_HOUR + sm
  const e = eh * MINUTES_PER_HOUR + em
  return t >= s && t <= e
}

/**
 * Löst Parameter-Platzhalter auf (severity-abhängig oder mit delta)
 */
function resolveParams(
  rule: Rule,
  sev: "ok" | "warn" | "crit",
  alert?: Alert
): Record<string, unknown> {
  const p = rule.params ?? {}
  const out: Record<string, unknown> = {}
  for (const [k, v] of Object.entries(p)) {
    if (
      v !== null &&
      typeof v === "object" &&
      "warn" in v &&
      "crit" in v
    ) {
      const obj = v as { warn?: unknown; crit?: unknown }
      out[k] = obj[sev] ?? obj.warn
    } else if (
      typeof v === "string" &&
      alert?.delta !== undefined &&
      v.includes("{delta}")
    ) {
      out[k] = v.replace("{delta}", String(alert.delta))
    } else {
      out[k] = v
    }
  }
  return out
}

/**
 * Policy-Entscheidungs-Engine
 * Matched Alert gegen Regeln und gibt Entscheidung zurück
 */
export function decide(userRoles: string[], alert: Alert, rules: Rule[]): Decision {
  const rule = rules.find(
    (r) => r.when.kpiId === alert.kpiId && r.when.severity.includes(alert.severity)
  )
  if (rule === undefined) {
    return { type: "deny", reason: "No matching rule" }
  }
  if (!withinWindow(rule.window)) {
    return { type: "deny", reason: "Outside window" }
  }

  const params = resolveParams(rule, alert.severity, alert)

  const appr = rule.approval
  const needsApproval =
    appr?.required === true &&
    !(appr.bypassIfSeverity !== undefined && alert.severity === appr.bypassIfSeverity)

  const approverRoles = appr?.roles

  const roleOk =
    !needsApproval || userRoles.some((r) => approverRoles?.includes(r as "admin" | "manager" | "operator"))

  const execute = rule.autoExecute === true && (!needsApproval || roleOk)

  return {
    type: "allow",
    execute,
    needsApproval,
    approverRoles,
    ruleId: rule.id,
    resolvedParams: params,
  }
}

File: app/policy/engine.py
"""
Policy Engine - Decision Logic
"""

from __future__ import annotations
from typing import List, Dict
from datetime import datetime
from .models import Alert, Decision, DecisionAllow, DecisionDeny, Rule, Role, Severity


def _within_window(window: dict | None, now: datetime | None = None) -> bool:
    """Prüft ob aktuelle Zeit im Zeitfenster liegt"""
    if not window:
        return True
    if now is None:
        now = datetime.now()

    day = now.weekday()  # 0=Mon .. 6=Sun
    days: List[int] = window.get("days", [])

    # Support für beide Konventionen: 0=Mon oder 0=Sun
    if day not in days and (day + 1) % 7 not in days:
        return False

    start = window.get("start", "00:00")
    end = window.get("end", "23:59")
    sh, sm = [int(x) for x in start.split(":")]
    eh, em = [int(x) for x in end.split(":")]
    t = now.hour * 60 + now.minute
    return (sh * 60 + sm) <= t <= (eh * 60 + em)


def _resolve_params(
    rule: Rule, sev: Severity, alert: Alert | None
) -> Dict[str, object]:
    """Löst Parameter-Platzhalter auf"""
    out: Dict[str, object] = {}
    for k, v in (rule.params or {}).items():
        if isinstance(v, dict) and "warn" in v and "crit" in v:
            # Severity-abhängige Werte
            out[k] = v.get(sev, v.get("warn"))
        elif (
            isinstance(v, str)
            and alert
            and ("{delta}" in v)
            and alert.delta is not None
        ):
            # Delta-Platzhalter
            out[k] = v.replace("{delta}", str(alert.delta))
        else:
            out[k] = v
    return out


def decide(user_roles: List[Role], alert: Alert, rules: List[Rule]) -> Decision:
    """
    Policy-Entscheidung treffen

    Args:
        user_roles: Rollen des aktuellen Users
        alert: Alert-Objekt
        rules: Liste aller Policies

    Returns:
        Decision (Allow oder Deny)
    """
    # Matching rule finden
    rule = next(
        (
            r
            for r in rules
            if r.when.kpiId == alert.kpiId and alert.severity in r.when.severity
        ),
        None,
    )

    if not rule:
        return DecisionDeny(type="deny", reason="No matching rule")

    # Zeitfenster prüfen
    if not _within_window(rule.window.model_dump() if rule.window else None):
        return DecisionDeny(type="deny", reason="Outside window")

    # Parameter auflösen
    params = _resolve_params(rule, alert.severity, alert)

    # Approval prüfen
    appr = rule.approval
    needs_approval = bool(
        appr
        and appr.required
        and not (appr.bypassIfSeverity and alert.severity == appr.bypassIfSeverity)
    )
    approver_roles = appr.roles if appr and appr.roles else None

    # Rollen-Check
    role_ok = (not needs_approval) or any(
        r in (approver_roles or []) for r in user_roles
    )
    execute = bool(rule.autoExecute and (not needs_approval or role_ok))

    return DecisionAllow(
        execute=execute,
        needsApproval=needs_approval,
        approverRoles=approver_roles,
        ruleId=rule.id,
        resolvedParams=params,
    )

